### ISSUE TITLE
Plot Decision regions by considering all dataset and all features

### ISSUE THREAD

**Issue by AlphaRandom** [rasbt/mlxtend#469]
Hi All,
I want to plot my decision regions by considering all features of my dataset for a MLP classifier.Looking at the example, it seems that I need to train my classifier by using all dataset and not just  a part of it as when I first train a classifier by using train_test_split function. So , the question is  : 
Do I need to train my classifier by considering all dataset,without first applying train_test_split function? Or May I train it by only using a part of the dataset and later applying the decision regions function?
Thanks

**Comment by AlphaRandom** [comment]
Well, I just had a look at the code of the decision_regions function and I found the answer. I have to train my classifier by using all dataset (all samples) , because it  later will predict by considering all new data samples got through the mesh grid. So there is no need to first apply the train_test_split function. If someone wants to confirm or not my reasoning is very welcome.

**Comment by rasbt** [comment]
Hi there,

You don't have to train your classifier on all examples in the dataset. If your dataset is already the training dataset, then of course, there is no further splitting necessary.

It really depends on what you want to show though. Btw have a look at example 4: http://rasbt.github.io/mlxtend/user_guide/plotting/plot_decision_regions/#example-4-highlighting-test-data-points

this shows how you can only fit on the training set for making the decision boundary but then highlight the test datapoints (which were not used for fitting) in the plot.

**Comment by AlphaRandom** [comment]
The example is clear, but my point is : I already have a dump of my trained classifier, in another script, based on X_train and  Y_train.I also got  X_test and Y_test  by using train_test_split function. If I wanted to use that dump to plot decision regions I couldn't use the X_test ,because I didn't save it on file. So , my question is : In this case should I first need to apply both train_test_split function and train my classifier later?

**Comment by AlphaRandom** [comment]
Regarding filler_feature_values and filler_feature_ranges parameters , I want to use average and standard deviation respectly, doubt is : Are these metrics to be calculated by considering the whole feature(all rows) or X_train (that contains a subset of rows for that feature) since classifier has been trained by X_train?

**Comment by AlphaRandom** [comment]
Third doubt, when I trained my MLP classifier I first used standarScaler() on my dataset . In your examples you directly pass X_train to train the classifier without any previous standardization . Why?

**Comment by rasbt** [comment]
> The example is clear, but my point is : I already have a dump of my trained classifier, in another script, based on X_train and Y_train.I also got X_test and Y_test by using train_test_split function. If I wanted to use that dump to plot decision regions I couldn't use the X_test ,because I didn't save it on file. So , my question is : In this case should I first need to apply both train_test_split function and train my classifier later?

Hm, I am not sure if I understand that correctly. You have saved the trained MLP classifiers (e.g., as pickle files) but not the test data? Actually, you don't need the test datapoints to show the decision regions of that classifier. You can also just show them for the training data or whatever dataset you have (the function does not require retraining the classifier). Unless you want to know where the test data points lie. 

> Regarding filler_feature_values and filler_feature_ranges parameters , I want to use average and standard deviation respectly, doubt is : Are these metrics to be calculated by considering the whole feature(all rows) or X_train (that contains a subset of rows for that feature) since classifier has been trained by X_train?

I would say it doesn't matter much in practice, because it is just for viewing a slice and does not influence the classifier itself. However, just according to convention, I would calculate that based on the training set only.


> Third doubt, when I trained my MLP classifier I first used standarScaler() on my dataset . In your examples you directly pass X_train to train the classifier without any previous standardization . Why?

If you are using an MLP, I would recommend using a StandardScaler like you did. The example in the documentation are more minimalistic, which is why i haven't used a standard scaling. In these examples, it doesn't matter much and I tried to keep the examples as short as possible for readability. 

So in your case, I would pass the output  of `yourscaler.transform(X_train)` to the `plot_decision_regions`, otherwise, you will get weird results.

**Comment by AlphaRandom** [comment]
> > The example is clear, but my point is : I already have a dump of my trained classifier, in another script, based on X_train and Y_train.I also got X_test and Y_test by using train_test_split function. If I wanted to use that dump to plot decision regions I couldn't use the X_test ,because I didn't save it on file. So , my question is : In this case should I first need to apply both train_test_split function and train my classifier later?
> 
> Hm, I am not sure if I understand that correctly. You have saved the trained MLP classifiers (e.g., as pickle files) but not the test data? Actually, you don't need the test datapoints to show the decision regions of that classifier. You can also just show them for the training data or whatever dataset you have (the function does not require retraining the classifier). Unless you want to know where the test data points lie.
Yes, I saved my pipeline object (serialized) ,after training,made of one standardScaler and the MLP classifier. In this case could I pass as X parameter this object to the function, after having it de-serialized ? Or Do I need to pass X_train(standardized) ?
> 
> > Regarding filler_feature_values and filler_feature_ranges parameters , I want to use average and standard deviation respectly, doubt is : Are these metrics to be calculated by considering the whole feature(all rows) or X_train (that contains a subset of rows for that feature) since classifier has been trained by X_train?
> 
> I would say it doesn't matter much in practice, because it is just for viewing a slice and does not influence the classifier itself. However, just according to convention, I would calculate that based on the training set only.
> 
> > Third doubt, when I trained my MLP classifier I first used standarScaler() on my dataset . In your examples you directly pass X_train to train the classifier without any previous standardization . Why?
> 
> If you are using an MLP, I would recommend using a StandardScaler like you did. The example in the documentation are more minimalistic, which is why i haven't used a standard scaling. In these examples, it doesn't matter much and I tried to keep the examples as short as possible for readability.
> 
> So in your case, I would pass the output of `yourscaler.transform(X_train)` to the `plot_decision_regions`, otherwise, you will get weird results.

OK, just to sum up,please tell me if this way to proceed is right :
 I'm starting from a X dataset , then I get X_train and X_test, then I will standardize them by getting X_train and X_test standardized
 (scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test). )
After that I will train my MLP classifier through X_train and   I will pass, as X parameter, to the decision_regions function X_train (standardized)  and not all X dataset,correct ? Other confirmation please : For filler_feature_values and filler_feature_ranges parameters ,I will pass two dictionaries by having for each excluded feature its average and standard deviation respectively ,always calculated by considering X_train (standardized) and not the whole X dataset (since average and standard deviation, for each column, would strongly  be different from those calculated from  X_train), correct?
Thanks in advance .

**Comment by rasbt** [comment]
> I'm starting from a X dataset , then I get X_train and X_test, then I will standardize them by getting X_train and X_test standardized
(scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test). )

That sounds good.

> After that I will train my MLP classifier through X_train and I will pass, as X parameter, to the decision_regions function X_train (standardized) and not all X dataset,correct ? 

Hm, note that the only thing that changes is the points that are overlaid with the decision region plot (the scatter points). You can either just pass X_train (standardized) to only show the training points. Or you can only pass X_test (standardized) to show the test points. The decision regions will not change though. If you want to show both training and test datapoints at the same time, you would pass a concatenated version of X_train (standardized) and X_test (standardized). But concatenating  X_train (standardized) and X_test (standardized) is the same as

X_train = scaler.fit(X_train)
X = scaler.transform(X)

Again, note that the decision region doesn't change, no matter what data you pass to the `plot_decision_regions` function. 


> Other confirmation please : For filler_feature_values and filler_feature_ranges parameters ,I will pass two dictionaries by having for each excluded feature its average and standard deviation respectively ,always calculated by considering X_train (standardized) and not the whole X dataset (since average and standard deviation, for each column, would strongly be different from those calculated from X_train), correct?
Thanks in advance .

I recommend using the mean and standard deviation from the training set (X_train) but it would *not* make a large difference whether it comes from the training set or the whole dataset in this case

**Comment by AlphaRandom** [comment]
> the points that are overlaid with the decision region plot



> > I'm starting from a X dataset , then I get X_train and X_test, then I will standardize them by getting X_train and X_test standardized
> > (scaler = StandardScaler()
> > X_train = scaler.fit_transform(X_train)
> > X_test = scaler.transform(X_test). )
> 
> That sounds good.
> 
> > After that I will train my MLP classifier through X_train and I will pass, as X parameter, to the decision_regions function X_train (standardized) and not all X dataset,correct ?
> 
> Hm, note that the only thing that changes is the points that are overlaid with the decision region plot (the scatter points). You can either just pass X_train (standardized) to only show the training points. Or you can only pass X_test (standardized) to show the test points. The decision regions will not change though. If you want to show both training and test datapoints at the same time, you would pass a concatenated version of X_train (standardized) and X_test (standardized). But concatenating X_train (standardized) and X_test (standardized) is the same as
> 
> X_train = scaler.fit(X_train)
> X = scaler.transform(X)

Yeah, it's something I was thinking to do like : (X_train and X_test is already meant standardized) 
np.concatenate((X_train,X_test),axis=0,out=X_new)
Then , retrieving average and standard deviation from X_new for each excluded feature to be used for the filler_feature_values and filler_feature_ranges parameters and to use always X_new as parameter of X in the function. But, I think that I will only use X_train (standardized) both as X parameter and to calculate the average and standard deviation of each feature to be put as filler_feature_values and filler_feature_ranges parameters , respectively. 
Thanks for your support!
> 
> Again, note that the decision region doesn't change, no matter what data you pass to the `plot_decision_regions` function.
> 
> > Other confirmation please : For filler_feature_values and filler_feature_ranges parameters ,I will pass two dictionaries by having for each excluded feature its average and standard deviation respectively ,always calculated by considering X_train (standardized) and not the whole X dataset (since average and standard deviation, for each column, would strongly be different from those calculated from X_train), correct?
> > Thanks in advance .
> 
> I recommend using the mean and standard deviation from the training set (X_train) but it would _not_ make a large difference whether it comes from the training set or the whole dataset in this case
